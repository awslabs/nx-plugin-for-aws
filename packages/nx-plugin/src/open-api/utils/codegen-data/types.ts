/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import type { Plugin } from '@hey-api/openapi-ts';
import type { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';

export type ClientData = Parameters<Plugin.LegacyHandler<any>>[0]['client'];
export type Operation = ClientData['services'][number]['operations'][number];
export type Model = ClientData['models'][number];

export interface CodeGenData extends ClientData {
  className: string;
  info: OpenAPIV3.InfoObject | OpenAPIV3_1.InfoObject;
  allOperations: Operation[];
  operationsByTag: { [tag: string]: Operation[] };
  untaggedOperations: Operation[];
  vendorExtensions: { [key: string]: any };
}

export const flattenModelLink = (link?: Model | Model[]): Model =>
  link === undefined ? undefined : Array.isArray(link) ? link[0] : link;

// Model types which indicate it is composed (ie inherits/mixin's another schema)
export const COMPOSED_SCHEMA_TYPES = new Set(['one-of', 'any-of', 'all-of']);
export const COLLECTION_TYPES = new Set(['array', 'dictionary']);
export const PRIMITIVE_TYPES = new Set([
  'string',
  'integer',
  'number',
  'boolean',
  'null',
  'any',
  'binary',
  'void',
]);
