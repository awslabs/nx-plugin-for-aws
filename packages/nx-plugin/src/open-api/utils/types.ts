/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import type { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';

export type Spec = OpenAPIV3.Document | OpenAPIV3_1.Document;
