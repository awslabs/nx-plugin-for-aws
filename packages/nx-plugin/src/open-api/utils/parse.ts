/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import SwaggerParser from '@apidevtools/swagger-parser';
import { Spec } from './types';

const TREE_PROTOCOL_PREFIX = 'workspace://';

/**
 * Parse an OpenAPI specification given its path within the tree
 */
export const parseOpenApiSpec = async (
  tree: Tree,
  specPath: string,
): Promise<Spec> => {
  // Use a custom protocol prefix to ensure that the swagger parser doesn't try and read from the filesystem itself
  const spec = await SwaggerParser.bundle(
    `${TREE_PROTOCOL_PREFIX}/${specPath}`,
    {
      resolve: {
        // Use a custom resolver which reads files from the tree rather than directly from the filesystem
        file: {
          order: 1,
          read: (file) => {
            const path = file.url.slice(TREE_PROTOCOL_PREFIX.length);
            if (!tree.exists(path)) {
              const message = `Unable to find ${specPath} in workspace ${tree.root}`;

              // Write an error to console as swagger parser's error message may be confusing with our custom protocol prefix
              console.error(message);
              throw new Error(message);
            }
            return tree.read(path);
          },
          canRead: (file) => file.url.startsWith(TREE_PROTOCOL_PREFIX),
        },
        http: false,
      },
    },
  );

  if ('swagger' in spec) {
    throw new Error(
      `OpenAPI v2 specifications are not supported. Please use OpenAPI v3`,
    );
  }
  return spec;
};
