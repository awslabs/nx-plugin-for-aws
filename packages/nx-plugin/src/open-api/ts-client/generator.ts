/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, Tree } from '@nx/devkit';
import { OpenApiTsClientGeneratorSchema } from './schema';
import { parseOpenApiSpec } from '../utils/parse';
import { buildOpenApiCodeGenData } from '../utils/codegen-data';
import * as path from 'path';
import { CodeGenData } from '../utils/codegen-data/types';
import { formatFilesInSubtree } from '../../utils/format';

/**
 * Generates typescript client from an openapi spec
 */
export const openApiTsClientGenerator = async (
  tree: Tree,
  options: OpenApiTsClientGeneratorSchema,
) => {
  const data = await buildOpenApiCodeGenerationData(
    tree,
    options.openApiSpecPath,
  );

  generateOpenApiTsClient(tree, data, options.outputPath);

  await formatFilesInSubtree(tree);
};

/**
 * Build a data structure which can be used to generate code from OpenAPI
 */
export const buildOpenApiCodeGenerationData = async (
  tree: Tree,
  specPath: string,
) => {
  const spec = await parseOpenApiSpec(tree, specPath);
  return await buildOpenApiCodeGenData(spec);
};

/**
 * Generate an OpenAPI typescript client in the target directory
 */
export const generateOpenApiTsClient = (
  tree: Tree,
  data: CodeGenData,
  outputPath: string,
) => {
  generateFiles(tree, path.join(__dirname, 'files'), outputPath, data);
};

export default openApiTsClientGenerator;
