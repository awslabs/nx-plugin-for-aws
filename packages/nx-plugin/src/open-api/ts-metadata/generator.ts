/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, Tree } from '@nx/devkit';
import { OpenApiMetadataSchema } from './schema';
import { parseOpenApiSpec } from '../utils/parse';
import { buildOpenApiCodeGenData } from '../utils/codegen-data';
import { formatFilesInSubtree } from '../../utils/format';
import path from 'path';
import { CodeGenData } from '../utils/codegen-data/types';

/**
 * Generates typescript metadata from an openapi spec
 */
export const openApiTsMetadataGenerator = async (
  tree: Tree,
  options: OpenApiMetadataSchema,
) => {
  const spec = await parseOpenApiSpec(tree, options.openApiSpecPath);

  const data = await buildOpenApiCodeGenData(spec);

  generateOpenApiTsMetadata(tree, data, options.outputPath);

  await formatFilesInSubtree(tree);
};

/**
 * Generate OpenAPI typescript metadata in the target directory
 */
export const generateOpenApiTsMetadata = (
  tree: Tree,
  data: CodeGenData,
  outputPath: string,
) => {
  generateFiles(tree, path.join(__dirname, 'files'), outputPath, data);
};

export default openApiTsMetadataGenerator;
