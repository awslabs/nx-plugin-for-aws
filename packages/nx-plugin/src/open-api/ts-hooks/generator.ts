/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { OpenApiHooksSchema } from './schema';
import { parseOpenApiSpec } from '../utils/parse';
import { buildOpenApiCodeGenData } from '../utils/codegen-data';
import { generateOpenApiTsClient } from '../ts-client/generator';

/**
 * Generates typescript hooks from an openapi spec
 */
export const openApiTsHooksGenerator = async (
  tree: Tree,
  options: OpenApiHooksSchema,
) => {
  const spec = await parseOpenApiSpec(tree, options.openApiSpecPath);

  const data = await buildOpenApiCodeGenData(spec);

  generateOpenApiTsClient(tree, data, options.outputPath);

  // TODO: generate hooks which wrap the client
};

export default openApiTsHooksGenerator;
