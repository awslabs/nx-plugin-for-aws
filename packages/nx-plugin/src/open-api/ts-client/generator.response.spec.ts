/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { openApiTsClientGenerator } from './generator';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import { Spec } from '../utils/types';
import {
  baseUrl,
  callGeneratedClient,
  expectTypeScriptToCompile,
} from './generator.utils.spec';
import { Tree } from '@nx/devkit';

describe('openApiTsClientGenerator - responses', () => {
  let tree: Tree;
  const title = 'TestApi';

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  const validateTypeScript = (paths: string[]) => {
    expectTypeScriptToCompile(tree, paths);
  };

  it('should handle default responses', async () => {
    const spec: Spec = {
      openapi: '3.0.0',
      info: { title, version: '1.0.0' },
      paths: {
        '/test': {
          get: {
            operationId: 'getTest',
            responses: {
              '200': {
                description: 'Success',
                content: {
                  'application/json': {
                    schema: {
                      type: 'object',
                      properties: {
                        message: { type: 'string' },
                      },
                      required: ['message'],
                    },
                  },
                },
              },
              default: {
                description: 'Error',
                content: {
                  'application/json': {
                    schema: {
                      type: 'object',
                      properties: {
                        error: { type: 'string' },
                      },
                      required: ['error'],
                    },
                  },
                },
              },
            },
          },
        },
      },
    };

    tree.write('openapi.json', JSON.stringify(spec));

    await openApiTsClientGenerator(tree, {
      openApiSpecPath: 'openapi.json',
      outputPath: 'src/generated',
    });

    validateTypeScript([
      'src/generated/client.gen.ts',
      'src/generated/types.gen.ts',
    ]);

    const types = tree.read('src/generated/types.gen.ts', 'utf-8');
    expect(types).toMatchSnapshot();

    const client = tree.read('src/generated/client.gen.ts', 'utf-8');
    expect(client).toMatchSnapshot();

    const mockFetch = vi.fn();

    // Test successful response
    mockFetch.mockResolvedValue({
      status: 200,
      json: vi.fn().mockResolvedValue({ message: 'Success' }),
    });
    expect(await callGeneratedClient(client, mockFetch, 'getTest')).toEqual({
      message: 'Success',
    });
    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/test`,
      expect.objectContaining({
        method: 'GET',
      }),
    );

    // Test default error response
    mockFetch.mockClear();
    mockFetch.mockResolvedValue({
      status: 500,
      json: vi.fn().mockResolvedValue({ error: 'Internal Server Error' }),
    });
    await expect(
      callGeneratedClient(client, mockFetch, 'getTest'),
    ).rejects.toThrow(
      expect.objectContaining({
        status: 500,
        error: { error: 'Internal Server Error' },
      }),
    );
    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/test`,
      expect.objectContaining({
        method: 'GET',
      }),
    );

    // Test default error response with another status code
    mockFetch.mockClear();
    mockFetch.mockResolvedValue({
      status: 404,
      json: vi.fn().mockResolvedValue({ error: 'Not found' }),
    });
    await expect(
      callGeneratedClient(client, mockFetch, 'getTest'),
    ).rejects.toThrow(
      expect.objectContaining({ status: 404, error: { error: 'Not found' } }),
    );
    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/test`,
      expect.objectContaining({
        method: 'GET',
      }),
    );
  });

  it('should handle only default response', async () => {
    const spec: Spec = {
      openapi: '3.0.0',
      info: { title, version: '1.0.0' },
      paths: {
        '/test': {
          get: {
            operationId: 'getTest',
            responses: {
              default: {
                description: 'Response',
                content: {
                  'application/json': {
                    schema: {
                      type: 'object',
                      properties: {
                        message: { type: 'string' },
                      },
                      required: ['message'],
                    },
                  },
                },
              },
            },
          },
        },
      },
    };

    tree.write('openapi.json', JSON.stringify(spec));

    await openApiTsClientGenerator(tree, {
      openApiSpecPath: 'openapi.json',
      outputPath: 'src/generated',
    });

    validateTypeScript([
      'src/generated/client.gen.ts',
      'src/generated/types.gen.ts',
    ]);

    const types = tree.read('src/generated/types.gen.ts', 'utf-8');
    expect(types).toMatchSnapshot();

    const client = tree.read('src/generated/client.gen.ts', 'utf-8');
    expect(client).toMatchSnapshot();

    const mockFetch = vi.fn();

    // Test successful response
    mockFetch.mockResolvedValue({
      status: 200,
      json: vi.fn().mockResolvedValue({ message: 'Success' }),
    });
    expect(await callGeneratedClient(client, mockFetch, 'getTest')).toEqual({
      message: 'Success',
    });
    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/test`,
      expect.objectContaining({
        method: 'GET',
      }),
    );

    // Test another response status code
    mockFetch.mockClear();
    mockFetch.mockResolvedValue({
      status: 400,
      json: vi.fn().mockResolvedValue({ message: 'Another' }),
    });
    expect(await callGeneratedClient(client, mockFetch, 'getTest')).toEqual({
      message: 'Another',
    });
    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/test`,
      expect.objectContaining({
        method: 'GET',
      }),
    );
  });

  it('should handle multiple response status codes', async () => {
    const spec: Spec = {
      openapi: '3.0.0',
      info: { title, version: '1.0.0' },
      paths: {
        '/test': {
          post: {
            operationId: 'testResponses',
            requestBody: {
              content: {
                'application/json': {
                  schema: {
                    type: 'object',
                    properties: {
                      data: { type: 'string' },
                    },
                  },
                },
              },
            },
            responses: {
              '2XX': {
                description: 'Success',
                content: {
                  'application/json': {
                    schema: {
                      type: 'object',
                      properties: {
                        result: { type: 'string' },
                      },
                    },
                  },
                },
              },
              '500': {
                description: 'Specific 500 error',
                content: {
                  'application/json': {
                    schema: {
                      type: 'object',
                      properties: {
                        reason: { type: 'string' },
                      },
                    },
                  },
                },
              },
              '5XX': {
                description: 'Internal Server Error',
                content: {
                  'application/json': {
                    schema: {
                      type: 'object',
                      properties: {
                        error: { type: 'string' },
                        trace: { type: 'string' },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    };

    tree.write('openapi.json', JSON.stringify(spec));

    await openApiTsClientGenerator(tree, {
      openApiSpecPath: 'openapi.json',
      outputPath: 'src/generated',
    });

    validateTypeScript([
      'src/generated/client.gen.ts',
      'src/generated/types.gen.ts',
    ]);

    const types = tree.read('src/generated/types.gen.ts', 'utf-8');
    expect(types).toMatchSnapshot();

    const client = tree.read('src/generated/client.gen.ts', 'utf-8');
    expect(client).toMatchSnapshot();

    const mockFetch = vi.fn();

    // Test 2XX responses as result
    mockFetch.mockResolvedValue({
      status: 200,
      json: vi.fn().mockResolvedValue({ result: 'Success' }),
    });
    expect(
      await callGeneratedClient(client, mockFetch, 'testResponses'),
    ).toEqual({ result: 'Success' });
    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/test`,
      expect.objectContaining({
        method: 'POST',
      }),
    );

    mockFetch.mockClear();
    mockFetch.mockResolvedValue({
      status: 201,
      json: vi.fn().mockResolvedValue({ result: 'Another Success' }),
    });
    expect(
      await callGeneratedClient(client, mockFetch, 'testResponses'),
    ).toEqual({ result: 'Another Success' });
    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/test`,
      expect.objectContaining({
        method: 'POST',
      }),
    );

    // Test explicit 500 error
    mockFetch.mockClear();
    mockFetch.mockResolvedValue({
      status: 500,
      json: vi.fn().mockResolvedValue({ reason: 'fail' }),
    });
    await expect(
      callGeneratedClient(client, mockFetch, 'testResponses'),
    ).rejects.toThrow(
      expect.objectContaining({ status: 500, error: { reason: 'fail' } }),
    );
    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/test`,
      expect.objectContaining({
        method: 'POST',
      }),
    );

    // Test other 5XX errors
    mockFetch.mockClear();
    mockFetch.mockResolvedValue({
      status: 503,
      json: vi.fn().mockResolvedValue({ error: 'fail' }),
    });
    await expect(
      callGeneratedClient(client, mockFetch, 'testResponses'),
    ).rejects.toThrow(
      expect.objectContaining({ status: 503, error: { error: 'fail' } }),
    );
    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/test`,
      expect.objectContaining({
        method: 'POST',
      }),
    );

    // Test other 5XX errors
    mockFetch.mockClear();
    mockFetch.mockResolvedValue({
      status: 504,
      json: vi.fn().mockResolvedValue({ error: 'fail 2' }),
    });
    await expect(
      callGeneratedClient(client, mockFetch, 'testResponses'),
    ).rejects.toThrow(
      expect.objectContaining({ status: 504, error: { error: 'fail 2' } }),
    );
    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/test`,
      expect.objectContaining({
        method: 'POST',
      }),
    );
  });
});
