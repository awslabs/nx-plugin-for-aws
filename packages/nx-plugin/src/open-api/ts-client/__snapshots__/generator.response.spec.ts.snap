// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`openApiTsClientGenerator - responses > should handle default responses 1`] = `
"export type GetTest200Response = {
  message: string;
};
export type GetTestdefaultResponse = {
  error: string;
};
export type GetTestDefaultError = {
  status: number;
  error: GetTestdefaultResponse;
};
export type GetTestError = GetTestDefaultError;
"
`;

exports[`openApiTsClientGenerator - responses > should handle default responses 2`] = `
"import type {
  GetTest200Response,
  GetTestdefaultResponse,
} from './types.gen.js';

/**
 * Utility for serialisation and deserialisation of API types.
 */
export class $IO {
  protected static $mapValues = (data: any, fn: (item: any) => any) =>
    Object.fromEntries(Object.entries(data).map(([k, v]) => [k, fn(v)]));

  public static GetTest200Response = {
    toJson: (model: GetTest200Response): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.message === undefined
          ? {}
          : {
              message: model.message,
            }),
      };
    },
    fromJson: (json: any): GetTest200Response => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        message: json['message'],
      };
    },
  };

  public static GetTestdefaultResponse = {
    toJson: (model: GetTestdefaultResponse): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.error === undefined
          ? {}
          : {
              error: model.error,
            }),
      };
    },
    fromJson: (json: any): GetTestdefaultResponse => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        error: json['error'],
      };
    },
  };
}

/**
 * Client configuration for TestApi
 */
export interface TestApiConfig {
  /**
   * Base URL for the API
   */
  url: string;
  /**
   * Custom instance of fetch. By default the global 'fetch' is used.
   * You can override this to add custom middleware for use cases such as adding authentication headers.
   */
  fetch?: typeof fetch;
  /**
   * Additional configuration
   */
  options?: {
    /**
     * By default, the client will add a Content-Type header, set to the media type defined for
     * the request in the OpenAPI specification.
     * Set this to false to omit this header.
     */
    omitContentTypeHeader?: boolean;
  };
}

/**
 * API Client for TestApi
 */
export class TestApi {
  private $config: TestApiConfig;

  constructor(config: TestApiConfig) {
    this.$config = config;

    this.getTest = this.getTest.bind(this);
  }

  private $url = (
    path: string,
    pathParameters: { [key: string]: any },
    queryParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): string => {
    const baseUrl = this.$config.url.endsWith('/')
      ? this.$config.url.slice(0, -1)
      : this.$config.url;
    const pathWithParameters = Object.entries(pathParameters).reduce(
      (withParams, [key, value]) =>
        withParams.replace(\`{\${key}}\`, encodeURIComponent(\`\${value}\`)),
      path,
    );
    const queryString = Object.entries(queryParameters)
      .map(([key, value]) => {
        if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
          return value
            .map(
              (v) => \`\${encodeURIComponent(key)}=\${encodeURIComponent(\`\${v}\`)}\`,
            )
            .join('&');
        }
        return \`\${encodeURIComponent(key)}=\${encodeURIComponent(Array.isArray(value) ? value.map(String).join(',') : String(value))}\`;
      })
      .join('&');
    return (
      baseUrl + pathWithParameters + (queryString ? \`?\${queryString}\` : '')
    );
  };

  private $headers = (
    headerParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): [string, string][] => {
    return Object.entries(headerParameters).flatMap(([key, value]) => {
      if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
        return value.map((v) => [key, String(v)]) as [string, string][];
      }
      return [[key, String(value)]];
    });
  };

  private $fetch: typeof fetch = (...args) =>
    (this.$config.fetch ?? fetch)(...args);

  public async getTest(): Promise<GetTest200Response> {
    const pathParameters: { [key: string]: any } = {};
    const queryParameters: { [key: string]: any } = {};
    const headerParameters: { [key: string]: any } = {};

    const body = undefined;

    const response = await this.$fetch(
      this.$url('/test', pathParameters, queryParameters),
      {
        headers: this.$headers(headerParameters),
        method: 'GET',
        body,
      },
    );

    if (response.status === 200) {
      return $IO.GetTest200Response.fromJson(await response.json());
    }
    throw {
      status: response.status,
      error: $IO.GetTestdefaultResponse.fromJson(await response.json()),
    };
  }
}
"
`;

exports[`openApiTsClientGenerator - responses > should handle multiple response status codes 1`] = `
"export type TestResponses2XXResponse = {
  result?: string;
};
export type TestResponses500Response = {
  reason?: string;
};
export type TestResponses5XXResponse = {
  error?: string;
  trace?: string;
};
export type TestResponsesRequestContent = {
  data?: string;
};

export type TestResponsesRequest = TestResponsesRequestContent | undefined;
export type TestResponses500Error = {
  status: 500;
  error: TestResponses500Response;
};
export type TestResponses5XXError = {
  status: _5XX;
  error: TestResponses5XXResponse;
};
export type TestResponsesError = TestResponses500Error | TestResponses5XXError;
export type _5XX =
  | 500
  | 501
  | 502
  | 503
  | 504
  | 505
  | 506
  | 507
  | 508
  | 509
  | 510
  | 511
  | 512
  | 513
  | 514
  | 515
  | 516
  | 517
  | 518
  | 519
  | 520
  | 521
  | 522
  | 523
  | 524
  | 525
  | 526
  | 527
  | 528
  | 529
  | 530
  | 531
  | 532
  | 533
  | 534
  | 535
  | 536
  | 537
  | 538
  | 539
  | 540
  | 541
  | 542
  | 543
  | 544
  | 545
  | 546
  | 547
  | 548
  | 549
  | 550
  | 551
  | 552
  | 553
  | 554
  | 555
  | 556
  | 557
  | 558
  | 559
  | 560
  | 561
  | 562
  | 563
  | 564
  | 565
  | 566
  | 567
  | 568
  | 569
  | 570
  | 571
  | 572
  | 573
  | 574
  | 575
  | 576
  | 577
  | 578
  | 579
  | 580
  | 581
  | 582
  | 583
  | 584
  | 585
  | 586
  | 587
  | 588
  | 589
  | 590
  | 591
  | 592
  | 593
  | 594
  | 595
  | 596
  | 597
  | 598
  | 599;
"
`;

exports[`openApiTsClientGenerator - responses > should handle multiple response status codes 2`] = `
"import type {
  TestResponses2XXResponse,
  TestResponses500Response,
  TestResponses5XXResponse,
  TestResponsesRequestContent,
  TestResponsesRequest,
} from './types.gen.js';

/**
 * Utility for serialisation and deserialisation of API types.
 */
export class $IO {
  protected static $mapValues = (data: any, fn: (item: any) => any) =>
    Object.fromEntries(Object.entries(data).map(([k, v]) => [k, fn(v)]));

  public static TestResponses2XXResponse = {
    toJson: (model: TestResponses2XXResponse): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.result === undefined
          ? {}
          : {
              result: model.result,
            }),
      };
    },
    fromJson: (json: any): TestResponses2XXResponse => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        ...(json['result'] === undefined
          ? {}
          : {
              result: json['result'],
            }),
      };
    },
  };

  public static TestResponses500Response = {
    toJson: (model: TestResponses500Response): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.reason === undefined
          ? {}
          : {
              reason: model.reason,
            }),
      };
    },
    fromJson: (json: any): TestResponses500Response => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        ...(json['reason'] === undefined
          ? {}
          : {
              reason: json['reason'],
            }),
      };
    },
  };

  public static TestResponses5XXResponse = {
    toJson: (model: TestResponses5XXResponse): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.error === undefined
          ? {}
          : {
              error: model.error,
            }),
        ...(model.trace === undefined
          ? {}
          : {
              trace: model.trace,
            }),
      };
    },
    fromJson: (json: any): TestResponses5XXResponse => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        ...(json['error'] === undefined
          ? {}
          : {
              error: json['error'],
            }),
        ...(json['trace'] === undefined
          ? {}
          : {
              trace: json['trace'],
            }),
      };
    },
  };

  public static TestResponsesRequestContent = {
    toJson: (model: TestResponsesRequestContent): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.data === undefined
          ? {}
          : {
              data: model.data,
            }),
      };
    },
    fromJson: (json: any): TestResponsesRequestContent => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        ...(json['data'] === undefined
          ? {}
          : {
              data: json['data'],
            }),
      };
    },
  };
}

/**
 * Client configuration for TestApi
 */
export interface TestApiConfig {
  /**
   * Base URL for the API
   */
  url: string;
  /**
   * Custom instance of fetch. By default the global 'fetch' is used.
   * You can override this to add custom middleware for use cases such as adding authentication headers.
   */
  fetch?: typeof fetch;
  /**
   * Additional configuration
   */
  options?: {
    /**
     * By default, the client will add a Content-Type header, set to the media type defined for
     * the request in the OpenAPI specification.
     * Set this to false to omit this header.
     */
    omitContentTypeHeader?: boolean;
  };
}

/**
 * API Client for TestApi
 */
export class TestApi {
  private $config: TestApiConfig;

  constructor(config: TestApiConfig) {
    this.$config = config;

    this.testResponses = this.testResponses.bind(this);
  }

  private $url = (
    path: string,
    pathParameters: { [key: string]: any },
    queryParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): string => {
    const baseUrl = this.$config.url.endsWith('/')
      ? this.$config.url.slice(0, -1)
      : this.$config.url;
    const pathWithParameters = Object.entries(pathParameters).reduce(
      (withParams, [key, value]) =>
        withParams.replace(\`{\${key}}\`, encodeURIComponent(\`\${value}\`)),
      path,
    );
    const queryString = Object.entries(queryParameters)
      .map(([key, value]) => {
        if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
          return value
            .map(
              (v) => \`\${encodeURIComponent(key)}=\${encodeURIComponent(\`\${v}\`)}\`,
            )
            .join('&');
        }
        return \`\${encodeURIComponent(key)}=\${encodeURIComponent(Array.isArray(value) ? value.map(String).join(',') : String(value))}\`;
      })
      .join('&');
    return (
      baseUrl + pathWithParameters + (queryString ? \`?\${queryString}\` : '')
    );
  };

  private $headers = (
    headerParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): [string, string][] => {
    return Object.entries(headerParameters).flatMap(([key, value]) => {
      if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
        return value.map((v) => [key, String(v)]) as [string, string][];
      }
      return [[key, String(value)]];
    });
  };

  private $fetch: typeof fetch = (...args) =>
    (this.$config.fetch ?? fetch)(...args);

  public async testResponses(
    input?: TestResponsesRequest,
  ): Promise<TestResponses2XXResponse> {
    const pathParameters: { [key: string]: any } = {};
    const queryParameters: { [key: string]: any } = {};
    const headerParameters: { [key: string]: any } = {};
    if (!this.$config.options?.omitContentTypeHeader) {
      headerParameters['Content-Type'] = 'application/json';
    }
    const body =
      input === undefined
        ? undefined
        : typeof input === 'object'
          ? JSON.stringify($IO.TestResponsesRequestContent.toJson(input))
          : String($IO.TestResponsesRequestContent.toJson(input));

    const response = await this.$fetch(
      this.$url('/test', pathParameters, queryParameters),
      {
        headers: this.$headers(headerParameters),
        method: 'POST',
        body,
      },
    );

    if (response.status === 500) {
      throw {
        status: response.status,
        error: $IO.TestResponses500Response.fromJson(await response.json()),
      };
    }
    if (response.status >= 200 && response.status < 300) {
      return $IO.TestResponses2XXResponse.fromJson(await response.json());
    }
    if (response.status >= 500 && response.status < 600) {
      throw {
        status: response.status,
        error: $IO.TestResponses5XXResponse.fromJson(await response.json()),
      };
    }
    throw new Error(
      \`Unknown response status \${response.status} returned by API\`,
    );
  }
}
"
`;

exports[`openApiTsClientGenerator - responses > should handle only default response 1`] = `
"export type GetTestdefaultResponse = {
  message: string;
};
export type GetTestError = never;
"
`;

exports[`openApiTsClientGenerator - responses > should handle only default response 2`] = `
"import type { GetTestdefaultResponse } from './types.gen.js';

/**
 * Utility for serialisation and deserialisation of API types.
 */
export class $IO {
  protected static $mapValues = (data: any, fn: (item: any) => any) =>
    Object.fromEntries(Object.entries(data).map(([k, v]) => [k, fn(v)]));

  public static GetTestdefaultResponse = {
    toJson: (model: GetTestdefaultResponse): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.message === undefined
          ? {}
          : {
              message: model.message,
            }),
      };
    },
    fromJson: (json: any): GetTestdefaultResponse => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        message: json['message'],
      };
    },
  };
}

/**
 * Client configuration for TestApi
 */
export interface TestApiConfig {
  /**
   * Base URL for the API
   */
  url: string;
  /**
   * Custom instance of fetch. By default the global 'fetch' is used.
   * You can override this to add custom middleware for use cases such as adding authentication headers.
   */
  fetch?: typeof fetch;
  /**
   * Additional configuration
   */
  options?: {
    /**
     * By default, the client will add a Content-Type header, set to the media type defined for
     * the request in the OpenAPI specification.
     * Set this to false to omit this header.
     */
    omitContentTypeHeader?: boolean;
  };
}

/**
 * API Client for TestApi
 */
export class TestApi {
  private $config: TestApiConfig;

  constructor(config: TestApiConfig) {
    this.$config = config;

    this.getTest = this.getTest.bind(this);
  }

  private $url = (
    path: string,
    pathParameters: { [key: string]: any },
    queryParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): string => {
    const baseUrl = this.$config.url.endsWith('/')
      ? this.$config.url.slice(0, -1)
      : this.$config.url;
    const pathWithParameters = Object.entries(pathParameters).reduce(
      (withParams, [key, value]) =>
        withParams.replace(\`{\${key}}\`, encodeURIComponent(\`\${value}\`)),
      path,
    );
    const queryString = Object.entries(queryParameters)
      .map(([key, value]) => {
        if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
          return value
            .map(
              (v) => \`\${encodeURIComponent(key)}=\${encodeURIComponent(\`\${v}\`)}\`,
            )
            .join('&');
        }
        return \`\${encodeURIComponent(key)}=\${encodeURIComponent(Array.isArray(value) ? value.map(String).join(',') : String(value))}\`;
      })
      .join('&');
    return (
      baseUrl + pathWithParameters + (queryString ? \`?\${queryString}\` : '')
    );
  };

  private $headers = (
    headerParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): [string, string][] => {
    return Object.entries(headerParameters).flatMap(([key, value]) => {
      if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
        return value.map((v) => [key, String(v)]) as [string, string][];
      }
      return [[key, String(value)]];
    });
  };

  private $fetch: typeof fetch = (...args) =>
    (this.$config.fetch ?? fetch)(...args);

  public async getTest(): Promise<GetTestdefaultResponse> {
    const pathParameters: { [key: string]: any } = {};
    const queryParameters: { [key: string]: any } = {};
    const headerParameters: { [key: string]: any } = {};

    const body = undefined;

    const response = await this.$fetch(
      this.$url('/test', pathParameters, queryParameters),
      {
        headers: this.$headers(headerParameters),
        method: 'GET',
        body,
      },
    );

    return $IO.GetTestdefaultResponse.fromJson(await response.json());
  }
}
"
`;
